lib/pbench/server/api/resources/visualize.py

```diff
@@ -0,0 +1,98 @@
+from http import HTTPStatus
+from urllib.request import Request
+
+from flask import current_app, jsonify
+from flask.wrappers import Response
+from pquisby.lib.post_processing import BenchmarkName, InputType, QuisbyProcessing
+
+from pbench.server import OperationCode, PbenchServerConfig
+from pbench.server.api.resources import (
+    APIAbort,
+    ApiAuthorizationType,
+    ApiBase,
+    ApiContext,
+    APIInternalError,
+    ApiMethod,
+    ApiParams,
+    ApiSchema,
+    Parameter,
+    ParamType,
+    Schema,
+)
+from pbench.server.cache_manager import (
+    CacheManager,
+    TarballNotFound,
+    TarballUnpackError,
+)
+from pbench.server.database import Dataset
+
+
+class Visualize(ApiBase):
+    """
+    API class to retrieve data using Quisby to visualize
+    """
+
+    def __init__(self, config: PbenchServerConfig):
+        super().__init__(
+            config,
+            ApiSchema(
+                ApiMethod.GET,
+                OperationCode.READ,
+                uri_schema=Schema(
+                    Parameter("dataset", ParamType.DATASET, required=True),
+                ),
+                authorization=ApiAuthorizationType.DATASET,
+            ),
+        )
+
+    def _get(
+        self, params: ApiParams, request: Request, context: ApiContext
+    ) -> Response:
+        """
+        This function is using Quisby to process results into a form that supports visualization
+
+        Args:
+            params: includes the uri parameters, which provide the dataset.
+            request: Original incoming Request object
+            context: API context dictionary
+
+        Raises:
+            APIAbort, reporting "NOT_FOUND" and "INTERNAL_SERVER_ERROR"
+
+        GET /api/v1/visualize/{dataset}
+        """
+
+        dataset = params.uri["dataset"]
+
+        cache_m = CacheManager(self.config, current_app.logger)
+        try:
+            tarball = cache_m.find_dataset(dataset.resource_id)
+        except TarballNotFound as e:
+            raise APIAbort(HTTPStatus.NOT_FOUND, str(e))
+
+        name = Dataset.stem(tarball.tarball_path)
+        try:
+            file = tarball.extract(tarball.tarball_path, f"{name}/result.csv")
+        except TarballUnpackError as e:
+            raise APIInternalError(str(e)) from e
```

Comment:
The "medium" (i.e., the result) has to have a `result.csv` file produced by the post-processing of a supported benchmark...otherwise, it's...um...unsupported.  😁

But, @dbutenhof, you concur that it should not be an internal error, right?

Reply:
Given that we've already failed if `dataset.metalog.pbench.script` isn't exactly "uperf", I'd be inclined to say by the time we get here we expect that this is a `pbench-uperf` benchmark wrapper execution which can reasonably be expected to have proper post-processing, and lack of the proper summary file is definitely unexpected. Maybe that's potentially more of an "agent-side internal error" but I can't see any reasonable way to express that in `HTTPStatus` aside from "internal server error". And, at best, I think we can reasonably say it means the server API was insufficiently paranoid about the dataset format...

Reply:
Perhaps I'm confused, but I think we're checking for the `result.csv` file before we're trying to decode the benchmark type.  (Do you think we should do this in a different order?)

Reply:
> where do we place "blame"?

Well, the client has made a request which the Server cannot fulfill based on the state of the data.  The Server is operating correctly, but it was given incomplete or inappropriately structured data by the client.  There is nothing which we can correct in the Server which would produce a successful result for the client, so I really don't think it's a server-side error of any sort (given that we're trying to become more liberal about the datasets that we accept).

Since `NOT_FOUND` isn't suitable (since the query on the dataset ID doesn't miss), and since we apparently don't like `UNSUPPORTED_MEDIA_TYPE` (even though that's arguably a good fit), we could go with:
- `405/METHOD_NOT_ALLOWED` - you cannot invoke `visualize` for this dataset (because it's missing the `result.csv` file)
- `422/UNPROCESSABLE_CONTENT` - this dataset cannot be visualized (because it's missing the `result.csv` file)
- Or, frankly, `400/BAD_REQUEST` is not unreasonable:  the request to visualize this dataset cannot be satisfied (because the dataset is missing the `result.csv` file).

I'm thinking we should reserve 418 until we really need it.

Reply:
`METHOD_NOT_ALLOWED` normally refers to `POST` vs `GET`/etc; I think that would be confusing when what we mean might best be respresented as "endpoint not allowed for this resource".

Yes, we've squeezed `UNSUPPORTED_MEDIA_TYPE`, but that's really intended to report incompatible request `Content-Type` (I presented XML but the API only supports JSON). I've never been entirely comfortable with this mapping, although at the time I hadn't noticed anything that looked measurably better or less misleading.

I'm not entirely happy with `UNPROCESSABLE_ENTITY`, either, which is oriented towards WebDAV protocol payloads ... and we're not talking about request payloads here... but as it's a niche error a lot less common than `UNSUPPORTED_MEDIA_TYPE`, it'd likely be a safer victim to steal and corrupt to our evil ends ...

Similarly for `BAD_REQUEST` ... this isn't a bad *request*, (there's nothing wrong with the URL, headers, or payload), but the resource isn't appropriate for the endpoint. You'd think this would be a common problem, but I really don't see any HTTP status with a description in that ballpark. It's even worse than trying to describe thread programming errors with those ridiculous UNIX error numbers... 😬 

I could actually see `501/NOT_IMPLEMENTED` for "benchmark other than uperf", come to think of it; along with the implicit indication that this might change in the future, which is wholly appropriate.

Interesting, although common usage (including ours) makes this potentially misleading, `403/FORBIDDEN` is technically described as "The request contained valid data and was understood by the server, but the server is refusing action." (Failed permission check is just a "for example".) Aside from that common inference, though, this isn't a bad fit at all. And the message explains *why* we're refusing action... 🤔

Reply:
Re `METHOD_NOT_ALLOWED`, Mozilla [explains it](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405) as
> indicates that the server knows the request method, but the target resource doesn't support this method.

which sounds pretty close to me (although, technically, we're supposed to respond with a list of methods which _are_ allowed, which sounds like a pain).

>  `UNSUPPORTED_MEDIA_TYPE`, but that's really intended to report incompatible request `Content-Type` 

Ah, yes, I'd missed the fact that it refers to the request payload.  Bummer.

> I'm not entirely happy with `UNPROCESSABLE_ENTITY`

Yeah, the WebDAV ones always look so attractive until you get to the details.  😛

Re `BAD_REQUEST`, Mozilla [describes it](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) as

> indicates that the server cannot or will not process the request due to something that is perceived to be a client error

I would suggest that the error is trying to visualize a dataset which has no `results.csv` file in it...the client just cannot to that.  😒

I like `501/NOT_IMPLEMENTED` in principal, but, based on what Mozilla [says](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501), it doesn't fit:

> 501 is the appropriate response when the server does not recognize the request method and is incapable of supporting it for any resource. The only methods that servers are required to support (and therefore that must not return 501) are [GET](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET) and [HEAD](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD).

That is, we do recognize the method (and, worse, it _is_ a `GET`).

Interestingly, Mozilla [goes on](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501) to suggest:
>If the server does recognize the method, but intentionally does not support it, the appropriate response is [405 Method Not Allowed](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405).

So, 405 sounds like a reasonable fit (better than 403, which I agree _would_ cover it, but too generically).

Reply:
But we *do* support the HTTP `POST` method for this endpoint. That's telling them they shouldn't be using a `POST` and is otherwise completely unhelpful. I think that's an incredibly bad fit. 😦 In particular, clients are encouraged to *cache* this to know that a subsequent `POST` would be inappropriate. It's not and can't be resource-specific.

Reply:
And, by the way, your Mozilla descriptions often conflict with the "easy reference" [Wikipedia](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes) page I often use, so from now on I'm sticking with [RFC 9110](https://datatracker.ietf.org/doc/html/rfc9110#name-status-codes) which differs from both in subtle ways that make them both liars.

Reply:
> from now on I'm sticking with [RFC 9110](https://datatracker.ietf.org/doc/html/rfc9110#name-status-codes)

Fair enough.  But, that reference says that [405 is resource-specific](https://datatracker.ietf.org/doc/html/rfc9110#name-405-method-not-allowed) (and therefore, hopefully, the caching is resource-sensitive, as well...).

Reply:
It would still mean rejecting a `POST` on the `/api/v1/datasets/<id>/visualize` with a quite specifically defined complaint that `POST` isn't acceptable and we only allow `POST`. And that's the problem; the HTTP errors are a very narrow and specifically defined set of definitions, which is absolutely ridiculous.

The most "wiggle room" I see in any of these is `FORBIDDEN` and `UNPROCESSABLE_ENTITY`, and I'm not thrilled about either of those.